### PR TITLE
feat(core): conditional built-in control flow

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Initial PoC of the built-in control flow support. The goal is to unblock compiler work.

### CODE REVIEW

**Review by crisbeto — APPROVED** [review]
LGTM. I left a few comments, but they're not blockers.

File: packages/core/src/render3/instructions/control_flow.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * built-in "if" and "switch". On the high level this instruction is responsible for adding and
+ * removing views selected by a conditional expression.
+ *
+ * @param containerIndex index of a container in a host view (indexed from HEADER_OFFSET) where
+ *     conditional views should be inserted.
+ * @param matchingTemplateIndex index of a template TNode representing a conditional view to be
+ *     inserted; -1 represents a special case when there is no view to insert.
+ * @codeGenApi
+ */
+export function ɵɵconditional<T>(containerIndex: number, matchingTemplateIndex: number, value?: T) {
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/render3/instructions/control_flow.ts:29]: After simplifications to the context (using the value as context) I don't think there is enough value in a separate instruction. We should be good with just one.

File: packages/core/src/render3/instructions/control_flow.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  const hostLView = getLView();
+  const bindingIndex = nextBindingIndex();
+  const lContainer = getLContainer(hostLView, HEADER_OFFSET + containerIndex);
+
+  if (bindingUpdated(hostLView, bindingIndex, matchingTemplateIndex)) {
+    // The index of the view to show changed - remove the previously displayed one
+    // (it is a noop if there are no active views in a container).
+    removeLViewFromLContainer(lContainer, 0);
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/render3/instructions/control_flow.ts:37]: Agreed, was going back and forth on this one. Will introduce an intermediate const for documentation purposes.

**Review by crisbeto — APPROVED** [review]
LGTM

**Review by alxhub — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]
👍

**Comment by pkozlowski-opensource** [comment]
caretaker note: G3 failures are unrelated

**Comment by pkozlowski-opensource** [comment]
This PR was merged into the repository by commit 93675dc797cb9f897c19fe298455dec52b900113.

**Pull request closed without merging**